hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java

```diff
@@ -0,0 +1,288 @@
+      public void onItemAdded(final CollectionItemEvent<? extends SourceT> event) {
+        inSync(new Runnable() {
+          @Override
+          public void run() {
+            myTokens.add(event.getIndex(), to.apply(event.getIndex(), event.getNewItem()));
+          }
+        });
+      }
+
+      @Override
+      public void onItemSet(final CollectionItemEvent<? extends SourceT> event) {
+        inSync(new Runnable() {
+          @Override
+          public void run() {
+            myTokens.set(event.getIndex(), to.apply(event.getIndex(), event.getNewItem()));
+          }
+        });
+      }
+
+      @Override
+      public void onItemRemoved(final CollectionItemEvent<? extends SourceT> event) {
+        inSync(new Runnable() {
+          @Override
+          public void run() {
+            myRemoveHandler.handle(event.getIndex());
+            myTokens.remove(event.getIndex());
+          }
+        });
+      }
+    });
+  }
+
+  @Override
+  public ModelT get() {
+    return myValue;
+  }
+
+  @Override
+  public ObservableList<Token> getTokens() {
+    return myTokens;
+  }
+
+  @Override
+  public String getPropExpr() {
+    return "PrettyHybridProperty[" + myTokens + "]";
+  }
+
+  @Override
+  public Registration addHandler(final EventHandler<? super PropertyChangeEvent<ModelT>> handler) {
+    if (myHandlers == null) {
+      myHandlers = new Listeners<>();
+    }
+    return myHandlers.add(handler);
+  }
+
+  private ModelT parse() {
+    return myParser.parse(myParsingContextFactory.getParsingContext(myTokens));
+  }
+
+  private void updateValue(ModelT newValue) {
+    final PropertyChangeEvent<ModelT> event = new PropertyChangeEvent<>(myValue, newValue);
+    myValue = newValue;
+
+    if (myHandlers != null) {
+      myHandlers.fire(new ListenerCaller<EventHandler<? super PropertyChangeEvent<ModelT>>>() {
+        @Override
+        public void call(EventHandler<? super PropertyChangeEvent<ModelT>> item) {
+          item.onEvent(event);
+        }
+      });
+    }
+  }
+
+  private void inSync(Runnable r) {
+    if (!myInSync) {
+      myInSync = true;
+      try {
+        r.run();
+      } finally {
+        myInSync = false;
+      }
+    }
+  }
+
+  private void sourceAdd(final int index, final Token item) {
+    myUpdateHandler.handle(new Runnable() {
+      @Override
+      public void run() {
+        mySource.add(index, myFrom.apply(index, item));
+      }
+    });
+  }
+
+  private void sourceSet(final int index, final Token newItem) {
+    myUpdateHandler.handle(new Runnable() {
+      @Override
+      public void run() {
+        mySource.set(index, myFrom.apply(index, newItem));
+      }
+    });
+  }
+
+  private void sourceRemove(final int index) {
+    myUpdateHandler.handle(new Runnable() {
+      @Override
+      public void run() {
+        mySource.remove(index);
+      }
+    });
+  }
+
+  public interface IndexedTransform <SourceT, TargetT> {
+    TargetT apply(int index, SourceT source);
```

Comment:
If implementation allocates some resources for source than this index could be used as an id for maintenance and dispose.